File: packages/scenarios/botcheck_scenarios/decision_slots.py
from __future__ import annotations

import re

DECISION_DEFAULT_SLOT = "default"
DECISION_PATH_SLOT_PREFIX = "path_"
DECISION_OUTPUT_HANDLE_PREFIX = "decision-output:"

_PATH_SLOT_RE = re.compile(rf"^{re.escape(DECISION_PATH_SLOT_PREFIX)}(\d+)$")


def decision_path_slot(index: int) -> str:
    safe_index = max(1, int(index))
    return f"{DECISION_PATH_SLOT_PREFIX}{safe_index}"


def decision_output_slots(output_count: int) -> list[str]:
    safe_count = max(1, int(output_count))
    slots = [DECISION_DEFAULT_SLOT]
    slots.extend(decision_path_slot(index) for index in range(1, safe_count))
    return slots


def decision_handle_id(slot: str) -> str:
    return f"{DECISION_OUTPUT_HANDLE_PREFIX}{slot}"


def parse_decision_handle_slot(handle_id: str | None) -> str | None:
    if not handle_id or not handle_id.startswith(DECISION_OUTPUT_HANDLE_PREFIX):
        return None
    slot = handle_id[len(DECISION_OUTPUT_HANDLE_PREFIX) :].strip().lower()
    return slot or None


def is_default_decision_slot(slot: str | None) -> bool:
    return (slot or "").strip().lower() == DECISION_DEFAULT_SLOT


def is_path_decision_slot(slot: str | None) -> bool:
    return _PATH_SLOT_RE.match((slot or "").strip().lower()) is not None


def decision_path_slot_index(slot: str | None) -> int | None:
    match = _PATH_SLOT_RE.match((slot or "").strip().lower())
    if not match:
        return None
    try:
        parsed = int(match.group(1))
    except (TypeError, ValueError):
        return None
    return parsed if parsed > 0 else None

File: web/src/lib/decision-slots.ts
export const DECISION_DEFAULT_SLOT = "default";
export const DECISION_PATH_SLOT_PREFIX = "path_";
export const DECISION_OUTPUT_HANDLE_PREFIX = "decision-output:";

const PATH_SLOT_PATTERN = new RegExp(`^${DECISION_PATH_SLOT_PREFIX}(\\d+)$`);

export function decisionPathSlot(indexRaw: number): string {
  const index = Math.max(1, Math.floor(indexRaw));
  return `${DECISION_PATH_SLOT_PREFIX}${index}`;
}

export function decisionOutputSlotsFromCount(outputCountRaw: number): string[] {
  const outputCount = Math.max(1, Math.floor(outputCountRaw));
  const slots: string[] = [DECISION_DEFAULT_SLOT];
  for (let index = 1; index < outputCount; index += 1) {
    slots.push(decisionPathSlot(index));
  }
  return slots;
}

export function decisionHandleId(slot: string): string {
  return `${DECISION_OUTPUT_HANDLE_PREFIX}${slot}`;
}

export function parseDecisionHandleSlot(handleId: string | null | undefined): string | null {
  if (!handleId || !handleId.startsWith(DECISION_OUTPUT_HANDLE_PREFIX)) {
    return null;
  }
  const slot = handleId.slice(DECISION_OUTPUT_HANDLE_PREFIX.length).trim().toLowerCase();
  return slot || null;
}

export function isDefaultDecisionSlot(slot: string | null | undefined): boolean {
  return (slot ?? "").trim().toLowerCase() === DECISION_DEFAULT_SLOT;
}

export function isPathDecisionSlot(slot: string | null | undefined): boolean {
  return PATH_SLOT_PATTERN.test((slot ?? "").trim().toLowerCase());
}

export function decisionPathSlotIndex(slot: string | null | undefined): number | null {
  const normalized = (slot ?? "").trim().toLowerCase();
  const match = normalized.match(PATH_SLOT_PATTERN);
  if (!match) {
    return null;
  }
  const parsed = Number.parseInt(match[1] ?? "", 10);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    return null;
  }
  return parsed;
}
